### PR TITLE
Support for sklearn 1.6 conformance testing

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

- scikit-learn-intelex PR to support internal conformance checks on sklearn 1.6 for CPU and 1.4 for GPU
- deselected tests updates to handle xfail/xpass discrepancies and a handful of outlying tests - all were investigated and large-scale failures have been fixed
- updates to daal4py logistic_path logic to align with scikit-learn
- updates to ridge to align coefficient and intercept shapes with scikit-learn logic

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

### CODE REVIEW

**Comment by ethanglaser** [comment]
http://intel-ci.intel.com/f02d307b-9927-f172-b696-a4bf010d0e2d

**Review by yuejiaointel — COMMENTED** [review]
Any impact of these extra checks on  throughput or memory?

File: daal4py/sklearn/linear_model/logistic_path.py [hunk 1]
```diff
@@ -753,7 +763,11 @@ def _func_(x, *args):
             else:
                 n_classes = max(2, classes.size)
                 if sklearn_check_version("1.1"):
-                    if solver in ["lbfgs", "newton-cg"]:
+                    if sklearn_check_version("1.6"):
+                        solver_list = ["lbfgs", "newton-cg", "newton-cholesky"]
```
Comment (icfaust) [inline daal4py/sklearn/linear_model/logistic_path.py:767]: Sklearn docs show newton-cholesky has been an option since 1.2. How come are we adding it for 1.6 (something to do with tests perhaps)?

File: deselected_tests.yaml [hunk 2]
```diff
@@ -444,3 +459,6 @@ gpu:
 
   # Introduced with RNG forest updates in oneDAL
   - ensemble/tests/test_voting.py::test_set_estimator_drop
+
+  # Fails from scikit-learn 1.4, SVC dependent (not supported on GPU)
+  - model_selection/tests/test_validation.py::test_learning_curve_some_failing_fits_warning
```
Comment (icfaust) [inline deselected_tests.yaml]: I had to also deselect this in #2494 for the new public GPU runner. I didn't put in any effort to figure out why.  Pretty curious, could you message me some of your investigation on this (so that we/I can write a ticket for it) and why this didn't pop up until now in private CI?

File: sklearnex/linear_model/ridge.py [hunk 3]
```diff
@@ -325,6 +325,15 @@ def _onedal_fit(self, X, y, sample_weight, queue=None):
             self._onedal_estimator.fit(X, y, queue=queue)
             self._save_attributes()
 
+            if sklearn_check_version("1.6"):
```
Comment (icfaust) [inline sklearnex/linear_model/ridge.py:328]: Could you do this for Linear Regression and the related incremental algos so that they all behave the same/ reduce maintenance?

**Review by icfaust — COMMENTED** [review]
Relatively straightforward, just some small questions/changes. I think we can get this easily ready for merge for when private CI is available again. The only actual change is to duplicate the Ridge changes to LinearRegression and the incrementals.

File: daal4py/sklearn/linear_model/logistic_path.py (hunk 1 above)
Comment (ethanglaser) [reply daal4py/sklearn/linear_model/logistic_path.py:767]: This commit from 8 months ago is why these changes were necessary: https://github.com/scikit-learn/scikit-learn/commit/c08b4332a3358f0090c8e3873aedde815908e248#diff-66422c7ed307653c80e5d68ed353d44d8fbb02319c428e140daff082eaf9d6cb. All the logic in this daal4py file is replicating sklearn _logistic.py. It's ugly. I created a ticket to look into how we can make this more flexible.

Comment (icfaust) [reply daal4py/sklearn/linear_model/logistic_path.py:767]: Yeah, totally agree. I think that is sufficient.

**Comment by icfaust** [comment]
@ethanglaser ping me directly once changes are made, so we can run private CI and merge this quickly.

File: sklearnex/linear_model/ridge.py (hunk 3 above)
Comment (ethanglaser) [reply sklearnex/linear_model/ridge.py:328]: I am not sure its relevant for the others, it originates from https://github.com/scikit-learn/scikit-learn/pull/19746 which specifically addresses Ridge

Comment (icfaust) [reply sklearnex/linear_model/ridge.py:328]: https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/linear_model/tests/test_common.py#L197 Is the answer hidden within the linked PR.  Place notes into LinearRegression and IncrementalLinearRegression about this deviation, and add this change in return values to IncrementalRidge.

File: daal4py/sklearn/linear_model/logistic_path.py [hunk 4]
```diff
@@ -359,14 +359,16 @@ def __logistic_regression_path(
             y_bin = np.ones(y.shape, dtype=X.dtype)
             # for compute_class_weight
 
-            if solver in ["lbfgs", "newton-cg"]:
+            if solver == "liblinear" or (
+                not sklearn_check_version("1.6") and solver not in ["lbfgs", "newton-cg"]
+            ):
+                mask_classes = np.array([-1, 1])
+                y_bin[~mask] = -1.0
```
Comment (david-cortes-intel) [inline daal4py/sklearn/linear_model/logistic_path.py:366]: Why would this be required if these are passed to sklearn? As far as I am aware, with solvers like 'liblinear' it still accepts inputs of zeros and ones. It's even a requirement for classifiers to be sklearn-compatible to be able to work with such inputs:
https://scikit-learn.org/stable/developers/develop.html#estimator-types

Does some test fail if sklearn is passed 'y' containing zeros and ones?

Comment (ethanglaser) [reply daal4py/sklearn/linear_model/logistic_path.py:366]: These changes are strictly for alignment with sklearn, see https://github.com/uxlfoundation/scikit-learn-intelex/pull/2465#discussion_r2122771629
